src/compiler/checker.ts

```diff
@@ -9003,7 +9006,7 @@ namespace ts {
                         //          breaking the intersection apart.
                         result = someTypeRelatedToType(<IntersectionType>source, target, /*reportErrors*/ false);
                     }
-                    if (!result && (source.flags & TypeFlags.StructuredOrTypeVariable || target.flags & TypeFlags.StructuredOrTypeVariable)) {
+                    if (!result && (source.flags & TypeFlags.StructuredOrTypeVariable || target.flags & TypeFlags.ObjectOrTypeVariable)) {

```

Comment:
Why don't we cover unions and intersections here anymore? Actually, a better question is: Why did we handle them before? I thought recursiveTypeRelatedTo was for object types only. Was this a mistake? Should it have been only intersections, but not unions?

Reply:
We should exclude unions and intersections on the target side because the result will always be false. It always was false previously, but took longer to compute.